### PR TITLE
Matching algorithm

### PR DESCRIPTION
**Pull request by Iyanuoluwa-abolade**
- Updated my schema.prisma with the necessary data for the matching algorithm
- Implemented the Gale-Shapley matching algorithm in the backend
- Created an backend endpoint for the matching algorithm
- Developed a UI and endpoint to implement the matching algorithm on the frontend and display the matching results

Image 7-16-24 at 5 35 PM
Image 7-16-24 at 5 35 PM
Image 7-16-24 at 5 35 PM (1)
Image 7-16-24 at 5 36 PM

### CODE REVIEW

File: backend/index.js [hunk 1]
```diff
@@ -4,15 +4,20 @@ import bodyParser from 'body-parser';
...
+import {galeShapley} from './matchAlgorithm.js';
+
 
 const app = express();
 const port = 3001;
 const YEAR_TO_MILLISECOND_CONVERSION_FACTOR = 365 * 24 * 60 * 60 * 1000
 env.config();
 
+
+
```
Comment (aliceyang121) [inline backend/index.js:20]: Remove extra lines

File: backend/index.js [hunk 2]
```diff
@@ -22,6 +27,8 @@ const sessionStore = new SequelizeStore({
     db: sequelize,
 });
 
+
+
```
Comment (aliceyang121) [inline backend/index.js:31]: same here

File: backend/prisma/migrations/20240711224536_init_preference_table/migration.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/*
+  Warnings:
+
+  - Added the required column `preference` to the `Adoptee` table without a default value. This is not possible if the table is not empty.
+  - Added the required column `preference` to the `Adopters` table without a default value. This is not possible if the table is not empty.
+  - Added the required column `status` to the `Adopters` table without a default value. This is not possible if the table is not empty.
```
Comment (aliceyang121) [inline backend/prisma/migrations/20240711224536_init_preference_table/migration.sql:6]: Address if needed!

File: backend/matchAlgorithm.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    for (const preference of preferences) {
+      const adopter = adopters.find((adopter) => adopter.id === preference.adopterId);
+      if (adopter && !matches[adopter.id]) {
+        matches[adopter.id] = adoptee;
+        break;
+      }
+    }
+  }
+  
+  return matches;
```
Comment (robintan98) [inline backend/matchAlgorithm.js:38]: Not sure where my last commit comment went, but this is not the correct implementation. Please reference this for the right pseudocode: https://www.geeksforgeeks.org/stable-marriage-problem/

Building off of what we worked on last Friday together, you will need to create the preference lists, then run a while loop to engage/free pairs until everyone has been engaged. Then (and only then) will this yield an optimal matching.

Comment (Iyanuoluwa-abolade) [reply backend/matchAlgorithm.js:38]: Noted. I will make use of this pseudocode for the right implementation

**Pull request merged**